App/Components/AsyncImage.js

```diff
@@ -22,59 +22,99 @@ export default class AsyncImage extends React.Component {
       this._createRequest()
       return false
     }
-    // request has been sent, loaded has become true
-    return this.state.requested && this.state.loaded !== nextState.loaded && nextState.loaded === true
+    return this.state.error !== nextState.error ||
+      this.state.loaded !== nextState.loaded ||
+      this.state.requested !== nextState.requested
   }
 
   componentWillUnmount () {
     this.hasCanceled_ = true
   }
 
+
   _createRequest () {
     this.setState(() => ({requested: true}))
     IPFS.getHashRequest(this.props.hash, this.props.path)
       .then(this._setSource)
-      .catch(() => { }) // todo: handle failed image requests vs. unmount
+      .catch(this._retry) // todo: handle failed hash requests vs. unmount
   }
 
   _retry () {
-    this.setState(() => ({retry: 0, loaded: false}))
-    IPFS.getHashRequest(this.props.hash, this.props.path)
-      .then(this._setSource)
-      .catch(() => { }) // todo: handle failed image requests vs. unmount
+    if (this.state.retry > 0 && !this.hasCanceled_) {
+      this.setState(() => ({retry: this.state.retry - 1, loaded: false}))
+      IPFS.getHashRequest(this.props.hash, this.props.path)
+        .then(this._setSource)
+        .catch(this._retry)
+    } else {
+      this._error()
+    }
+  }
+
+  _error () {
+    // Calls if the image or hash fails to load
+    this.setState(() => ({error: true, loaded: false}))
+  }
+
+  _success () {
+    // Calls after image succesfully loads. Resets available retries for later loads
+    this.setState(() => ({retry: 2}))
   }
 
   _setSource = (source) => {
+    // After async token is received, it readys the image source
     if (!this.hasCanceled_) {
       this.setState(() => ({loaded: true, source}))
     }
   }
 
+  placeholder () {
+    if (this.state.error) {
```

Comment:
Lots or repeated code in here. I'd just create a `let sourceFile` before the `if/else if` statements, the the value of `sourceFile` in the `if/if else` then just have the `<Image>` code once, and it uses `sourceFile`.

Reply:
nice